File: WondayWall.iOS/WondayWall/Models/PromptContext.swift
import Foundation

// 画像生成プロンプト用カレンダーイベント
struct PromptCalendarEvent: Encodable {
    let id: String
    let title: String
    // "today", "tomorrow", "in N days" などの近接タグ
    let proximityTag: String
    let startTime: Date
    let endTime: Date?
    let location: String?
    let description: String?
}

// 画像生成プロンプト用ニューストピック（ランタイム専用、JSON 保存なし）
struct PromptNewsTopic: Encodable {
    let id: String
    let title: String
    let summary: String?
    let url: String?
    let publishedAt: Date?
    let ogpImageUrl: String?
    // OGP 画像データ（ランタイム専用。Gemini API の JSON には含めない）
    var ogpImageData: Data? = nil
    var ogpImageMimeType: String? = nil

    // ogpImageData / ogpImageMimeType を Encodable から除外する
    enum CodingKeys: String, CodingKey {
        case id, title, summary, url, publishedAt, ogpImageUrl
    }
}

// 画像生成に渡すコンテキスト一式（ランタイム専用、JSON 保存なし）
struct PromptContext {
    var calendarEvents: [PromptCalendarEvent] = []
    var newsTopics: [PromptNewsTopic] = []
    // Gemini API で利用可能な縦長比率（imageSize は "512" / "1K" / "2K" / "4K" 形式）
    var imageSize: String = "2K"
    var aspectRatio: String = "9:16"
    var additionalConstraints: String = ""
    // ベース壁紙のファイルパス（前回生成画像を参照する場合）
    var baseImagePath: String?
}

// コンテキスト構築の結果
struct ContextBuildResult {
    let promptContext: PromptContext
    let calendarEvents: [CalendarEventItem]
    let newsTopics: [NewsTopicItem]
}

File: WondayWall.Android/app/src/main/kotlin/com/studiofreesia/wondaywall/models/PromptContext.kt
package com.studiofreesia.wondaywall.models

import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

// 画像生成 API に渡すカレンダーイベント（Gemini API プロンプト用）
@Serializable
data class PromptCalendarEvent(
    val id: String,
    val title: String,
    // イベントの近接タグ（"today" / "tomorrow" / "in N days"）
    val proximityTag: String,
    val startTime: Instant,
    val endTime: Instant?,
    val location: String?,
    val description: String?,
)

// 画像生成 API に渡すニューストピック（Gemini API プロンプト用）
@Serializable
data class PromptNewsTopic(
    val id: String,
    val title: String,
    val summary: String?,
    val url: String?,
    val publishedAt: Instant?,
    val ogpImageUrl: String?,
    // 画像生成 API に添付する OGP 画像データ。プロンプト用 JSON には含めない。
    @Transient
    val ogpImageData: ByteArray? = null,
    @Transient
    val ogpImageMimeType: String? = null,
)

// 画像生成に渡すコンテキスト一式
data class PromptContext(
    // カレンダーイベント候補（7日以内の今後の予定）
    val calendarEvents: List<PromptCalendarEvent> = emptyList(),
    // ニューストピック候補（最大10件）
    val newsTopics: List<PromptNewsTopic> = emptyList(),
    // 画像サイズ（"2K"）
    val imageSize: String = "2K",
    // アスペクト比（Android向け縦長 "9:16"）
    val aspectRatio: String = "9:16",
    // ユーザーの追加制約・プロンプト
    val additionalConstraints: String = "",
)

// コンテキスト構築の結果
data class ContextBuildResult(
    val promptContext: PromptContext,
    val calendarEvents: List<CalendarEventItem>,
    val newsTopics: List<NewsTopicItem>,
)
